### PR TITLE
Fuzzing with crowbar

### PR DESCRIPTION
**Pull request by icristescu**
Added Crowbar tests can be run either as simple tests (random number generators) or using `afl-fuzz`.  

Added a fix in `index.ml`: when clearing an index, the corresponding index file is closed.

### CODE REVIEW

File: fuzz/main.ml [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+module Index = Index_unix.Make (Key) (Value)
+
+let t = Index.v ~fresh:true ~log_size index_name
+
+let string_gen size =
+  map [ bytes ] (fun s ->
+      try String.sub s 0 size with Invalid_argument _ -> bad_test ())
+
+let list_gen =
```
Comment (pascutto) [inline fuzz/main.ml]: I think you can also use `Crowbar.list1` here.

File: fuzz/main.ml [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+let find_present l t =
+  let ok = List.for_all (fun (k, v) -> List.mem v (Index.find_all t k)) l in
+  Index.clear t;
+  ok
+
+let no_extra_values_added l t =
+  Index.iter
+    (fun k v ->
+      if not (List.exists (fun (k', v') -> k = k' && v = v') l) then
```
Comment (pascutto) [inline fuzz/main.ml]: Is there a reason for not using `List.mem` here ?

File: fuzz/main.ml [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+  add_test ~name:"find_present_lwt" [ list_gen ] (fun l ->
+      check (Lwt_main.run (find_present_lwt l t)))
+
+let restart_tests () =
+  let w = Index.v ~fresh:false ~log_size index_name in
+  add_test ~name:"on restart: find_present" [ list_gen ] (fun l ->
+      check (find_present l w));
+  add_test ~name:"on restart: no_extra_values" [ list_gen ] (fun l ->
+      check (no_extra_values_added l w));
+  add_test ~name:"on restart:add value and duplicate key"
```
Comment (pascutto) [inline fuzz/main.ml]:
```suggestion
  add_test ~name:"on restart: add value and duplicate key"
```

File: fuzz/main.ml [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+    [ list_gen; string_gen value_size ] (fun l v ->
+      guard (l != []);
+      check (test_add l v w))
+
+let readonly_test () =
+  let r =
+    Index.v ~shared:true ~fresh:false ~readonly:true ~log_size index_name
```
Comment (pascutto) [inline fuzz/main.ml]: That `shared:true` with `readonly:true` seems wrong to me.
Merging #34 beforehand would solve probably solve this API ambiguity.

File: fuzz/main.ml [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+  let pp s = Fmt.fmt "%s" s
+end
+
+module Index = Index_unix.Make (Key) (Value)
+
+let t = Index.v ~fresh:true ~log_size index_name
+
+let string_gen size =
+  map [ bytes ] (fun s ->
+      try String.sub s 0 size with Invalid_argument _ -> bad_test ())
```
[the remark this answers is not in the document]
Comment (icristescu) [reply fuzz/main.ml]: `Crowbar.byes_fixed` is part of the not yet released version of `Crowbar`. I'll add a comment to the code to remember to update once a new release of `Crowbar` is available.

File: fuzz/main.ml (hunk 4 above)
Comment (pascutto) [reply fuzz/main.ml]: #34 has been merged.

**Comment by CraigFe** [comment]
Time to mercilessly kill this PR.

**Pull request closed without merging**